.github/workflows/pythonpublish.yml

```diff
@@ -25,15 +23,35 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          pip install setuptools build twine
+          pip install setuptools build
 
       - name: Build
         run: |
           python -m build
 
-      - name: Publish
-        env:
-          TWINE_USERNAME: __token__
-          TWINE_PASSWORD: ${{ secrets.PYPI_TOKEN }}
-        run: |
-          twine upload dist/*
+      - name: "Upload dists"
+        uses: "actions/upload-artifact@v4"
+        with:
+          name: "dist"
+          path: "dist/"
+          if-no-files-found: error
+          retention-days: 5
+
+  publish:
+    name: Upload release to PyPI
+    runs-on: ubuntu-latest
+    needs: "build"
+    environment:
+      name: release
+      url: https://pypi.org/p/hass-nabucasa
+    permissions:
+      id-token: write
+    steps:
+    - name: "Download dists"
+      uses: "actions/download-artifact@v4"
+      with:
+        name: "dist"
+        path: "dist/"
+
+    - name: "Publish dists to PyPI"
+      uses: "pypa/gh-action-pypi-publish@e53eb8b103ffcb59469888563dc324e3c8ba6f06" # v1.8.12
```

Comment:
Is the commit pin a security measure?

Reply:
Yes, this is the action that is responsible for generating a token, and uploading the files to pypi.

Tags/releases can be replaced, commits can not.